src/effects/effects_base/internal/effectsprovider.cpp

```diff
@@ -182,3 +187,241 @@ muse::Ret EffectsProvider::performEffect(AudacityProject& project, Effect* effec
+    }
+
+    if (t1 <= mT0) {
+        return;
+    }
+
+    bool success = true;
+
+    auto cleanup = finally([&] {
+        // Effect is already inited; we will call Process and then Init
+        // again, so the state is exactly the way it was before Preview
+        // was called.
+        if (!dryOnly) {
+            // TODO remove this reinitialization of state within the Effect object
+            // It is done indirectly via Effect::Instance
+            if (auto pInstance
+                    =std::dynamic_pointer_cast<EffectInstanceEx>(effect.MakeInstance())
+                    ) {
+                pInstance->Init();
+            }
+        }
+    });
+
+    auto vr0 = valueRestorer(mT0);
+    auto vr1 = valueRestorer(mT1);
+    // Most effects should stop at t1.
+    if (!previewFullSelection) {
+        mT1 = t1;
+    }
+
+    // In case any dialog control depends on mT1 or mDuration:
+    if (updateUI) {
+        updateUI();
+    }
+
+    // Save the original track list
+    auto saveTracks = mTracks;
+
+    auto cleanup2 = finally([&] {
+        mTracks = saveTracks;
+        if (*FocusDialog) {
+            BasicUI::SetFocus(*FocusDialog);
+        }
+    });
+
+    // Build NEW tracklist from rendering tracks
+    // Set the same owning project, so FindProject() can see it within Process()
+    const auto pProject = saveTracks->GetOwner();
+    mTracks = TrackList::Create(pProject);
+
+    // Linear Effect preview optimised by pre-mixing to one track.
+    // Generators need to generate per track.
+    if (isLinearEffect && !isGenerator) {
+        auto newTrack = MixAndRender(
+            saveTracks->Selected<const WaveTrack>(),
+            Mixer::WarpOptions { saveTracks->GetOwner() },
+            wxString {}, // Don't care about the name of the temporary tracks
+            factory, rate, floatSample, mT0, t1);
+        if (!newTrack) {
+            return;
+        }
+        mTracks->Add(newTrack);
+
+        newTrack->MoveTo(0);
+        newTrack->SetSelected(true);
+    } else {
+        for (auto src : saveTracks->Selected<const WaveTrack>()) {
+            auto dest = src->Copy(mT0, t1);
+            dest->SetSelected(true);
+            mTracks->Add(dest);
+        }
+    }
+
+    // NEW tracks start at time zero.
+    // Adjust mT0 and mT1 to be the times to process, and to
+    // play back in these tracks
+    mT1 -= mT0;
+    mT0 = 0.0;
+
+    // Update track/group counts
+    effect.CountWaveTracks();
+
+    // Apply effect
+    if (!dryOnly) {
+        using namespace BasicUI;
+        auto progress = MakeProgress(
+            effect.GetName(),
+            XO("Preparing preview"),
+            ProgressShowStop
+            ); // Have only "Stop" button.
+        auto vr = valueRestorer(mProgress, progress.get());
+
+        auto vr2 = valueRestorer(mIsPreview, true);
+
+        access.ModifySettings([&](EffectSettings& settings){
+            // Preview of non-realtime effect
+            auto pInstance
```

Comment:
If dynamic_pointer_cast fails, pInstance will be nullptr, leading to undefined behavior if pInstance->Init() is called on a null pointer. Always check whether the cast was successful before using the pointer.

Reply:
This is "Added effect preview function as is" - i.e. it is just a copy of the code from AU3 as is. 
The next step will be refactoring.
Done in several steps to see what will change

Reply:
> If dynamic_pointer_cast fails, pInstance will be nullptr, leading to undefined behavior if pInstance->Init() is called on a null pointer. Always check whether the cast was successful before using the pointer.

There is a check in this case